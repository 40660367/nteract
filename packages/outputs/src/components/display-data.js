// @flow strict
import * as React from "react";

import { RichMedia } from "./rich-media";

import type { DisplayDataOutput } from "@nteract/records";
type Props = DisplayDataOutput;

export const DisplayData = (props: Props) => {
  const { data, metadata, children } = props;

  return (
    <RichMedia data={data} metadata={metadata}>
      {children}
    </RichMedia>
  );
};

DisplayData.defaultProps = {
  outputType: "display_data",
  data: {},
  metadata: {}
};
