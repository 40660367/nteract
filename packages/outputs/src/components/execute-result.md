The `ExecuteResult` component is a container component that is used to render output data from the kernel. Similar to the `DisplayData` component, the `ExecuteResult` component acts as a simple switch over its children and correctly associates the media type of the data with the media type that a component supports.

Why do so many component employ this pattern? It's in the interest of readability. Although each component executes the same logic under the hood, the separation of components by output types allows you to write meaningful JSX from these React components.

```jsx
const Plain = props => <pre>{props.data}</pre>;
Plain.defaultProps = {
  mediaType: "text/plain"
};

<ExecuteResult data={{ "text/plain": "The answer to everything is 42." }}>
  <Plain />
</ExecuteResult>;
```
