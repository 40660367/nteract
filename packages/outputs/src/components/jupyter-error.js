// @flow strict
import * as React from "react";
import Ansi from "ansi-to-react";

import type { ErrorOutput } from "@nteract/records";
type Props = ErrorOutput;

export const JupyterError = (props: Props) => {
  const { ename, evalue, traceback } = props;

  if (!Array.isArray(traceback) || !traceback.length) {
    return (
      <Ansi className="nteract-display-area-traceback">{`${ename}: ${evalue}`}</Ansi>
    );
  }

  return (
    <Ansi className="nteract-display-area-traceback">
      {traceback.join("\n")}
    </Ansi>
  );
};

JupyterError.defaultProps = {
  outputType: "error",
  ename: "",
  evalue: "",
  traceback: []
};
