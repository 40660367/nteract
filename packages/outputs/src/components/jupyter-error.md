The `<JuptyerError />` component will render a Jupyter error with or without a traceback.

```jsx
<JupyterError ename="NameError" evalue="Yikes!" traceback={["Yikes, never heard of that one!"]} />
```

```jsx
<JupyterError ename="NameError" evalue="Yikes!" />
```