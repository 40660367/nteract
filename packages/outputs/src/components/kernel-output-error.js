// @flow strict
import * as React from "react";
// $FlowFixMe
import Ansi from "ansi-to-react";

type Props = {
  /**
   *  The name of the exception. This value is returned by the kernel.
   */
  ename: string,
  /**
   * The value of the exception. This value is returned by the kernel.
   */
  evalue: string,
  /**
   * The output type passed to the Output component. This should be `error`
   * if you would like to render a KernelOutputError component.
   */
  outputType: string,
  /**
   * The tracebook of the exception. This value is returned by the kernel.
   */
  traceback: Array<string>
};

export const KernelOutputError = (props: Props) => {
  const { ename, evalue, traceback } = props;

  if (!Array.isArray(traceback) || !traceback.length) {
    return <Ansi>{`${ename}: ${evalue}`}</Ansi>;
  }

  return <Ansi>{traceback.join("\n")}</Ansi>;
};

KernelOutputError.defaultProps = {
  outputType: "error",
  ename: "",
  evalue: "",
  traceback: []
};
