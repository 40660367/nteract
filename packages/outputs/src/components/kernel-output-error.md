The `<KernelOutputError />` component will render a Jupyter error with or without a traceback.

Per the Jupyter Messaging Protocol, the kernel will return the following payload if it encountered an error during the execution of a code cell.

```json
{
   'status' : 'error',
   'ename' : string,
   'evalue' : string,
   'traceback' : string[],
}
```

The `ename` property contains the name of the error (`OutOfMemory` or `NameError`). The `evalue` property contains the value of the error. This is usually the first line you see before the complete tracebook. The `traceback` property contains the stacktrace of the error as a list of strings, which each string consisting of a line within the stacktrace.

We provide a special `KernelOutputError` component that will allow you to render the error payloads received from the kernel. You can pass the `ename,` `evalue`, and `tracebook` property values to render the error and tracebook in a clean format.

```jsx
<KernelOutputError ename="NameError" evalue="Yikes!" traceback={["Yikes, never heard of that one!"]} />
```

You can also just render the `ename` and `evalue` for a more simplified view of error messages.

```jsx
<KernelOutputError ename="NameError" evalue="Yikes!" />
```