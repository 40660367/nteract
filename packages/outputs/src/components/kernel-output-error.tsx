import Ansi from "ansi-to-react";
import * as React from "react";
import styled from "styled-components";

interface Props {
  /**
   *  The name of the exception. This value is returned by the kernel.
   */
  ename: string;
  /**
   * The value of the exception. This value is returned by the kernel.
   */
  evalue: string;
  /**
   * The output type passed to the Output component. This should be `error`
   * if you would like to render a KernelOutputError component.
   */
  output_type: "error";
  /**
   * The tracebook of the exception. This value is returned by the kernel.
   */
  traceback: string[];
}

const PlainKernelOutputError = (props: Partial<Props>) => {
  const { ename, evalue, traceback } = props;

  const joinedTraceback = Array.isArray(traceback)
    ? traceback.join("\n")
    : traceback;

  const kernelOutputError = [];

  if (joinedTraceback) {
    kernelOutputError.push(joinedTraceback);
  } else {
    if (ename && evalue) {
      kernelOutputError.push(`${ename}: ${evalue}`);
    }
  }

  return <Ansi linkify={false}>{kernelOutputError.join("\n")}</Ansi>;
};

export const KernelOutputError = styled(PlainKernelOutputError)`
  & code {
    white-space: pre-wrap;
  }
`;

KernelOutputError.defaultProps = {
  output_type: "error",
  ename: "",
  evalue: "",
  traceback: []
};

KernelOutputError.displayName = "KernelOutputError";
