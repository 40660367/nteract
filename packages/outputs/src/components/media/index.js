export { HTML } from "./html";
export { Json } from "./json";
export { SVG } from "./svg";
export { Plain } from "./plain";
export { Markdown } from "./markdown";
export { Image } from "./image";
