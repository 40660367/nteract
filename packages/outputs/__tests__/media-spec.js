import React from "react";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";

import * as Media from "../src/components/media";

describe("HTML", () => {
  it("renders direct HTML", () => {
    const component = mount(<Media.HTML data={"<b>cats are best</b>"} />);
    expect(toJson(component)).toMatchSnapshot();
  });
  it("updates the underlying HTML when data changes", () => {
    const component = mount(<Media.HTML data={"<b>cats are best</b>"} />);
    component.setProps({ data: "<b>squirrels are pretty great, too</b>" });
    expect(toJson(component)).toMatchSnapshot();
  });
});
