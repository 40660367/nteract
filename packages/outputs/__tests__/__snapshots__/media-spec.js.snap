// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML renders direct HTML 1`] = `
<HTML
  data="<b>cats are best</b>"
  mediaType="text/html"
>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<b>cats are best</b>",
      }
    }
  />
</HTML>
`;

exports[`HTML updates the underlying HTML when data changes 1`] = `
<HTML
  data="<b>squirrels are pretty great, too</b>"
  mediaType="text/html"
>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<b>squirrels are pretty great, too</b>",
      }
    }
  />
</HTML>
`;
